src/server/pkg/metrics/metrics.go

```diff
@@ -12,53 +13,131 @@ import (
+func IncrementUserAction(userID string, action string) {
+	incrementActionChannel <- &incrementUserAction{
+		action: action,
+		user:   userID,
+	}
+}
+
+func (r *Reporter) dbMetrics(metrics *Metrics) {
 	cursor, err := gorethink.Object(
 		"Repos",
-		gorethink.DB(pfsDbName).Table("Repos").Count(),
+		gorethink.DB(r.pfsDbName).Table("Repos").Count(),
 		"Commits",
-		gorethink.DB(pfsDbName).Table("Commits").Count(),
+		gorethink.DB(r.pfsDbName).Table("Commits").Count(),
 		"ArchivedCommits",
-		gorethink.DB(pfsDbName).Table("Commits").Filter(
+		gorethink.DB(r.pfsDbName).Table("Commits").Filter(
 			map[string]interface{}{
 				"Archived": true,
 			},
 		).Count(),
 		"CancelledCommits",
-		gorethink.DB(pfsDbName).Table("Commits").Filter(
+		gorethink.DB(r.pfsDbName).Table("Commits").Filter(
 			map[string]interface{}{
 				"Cancelled": true,
 			},
 		).Count(),
 		"Files",
-		gorethink.DB(pfsDbName).Table("Diffs").Group("Path").Ungroup().Count(),
+		gorethink.DB(r.pfsDbName).Table("Diffs").Group("Path").Ungroup().Count(),
 		"Jobs",
-		gorethink.DB(ppsDbName).Table("JobInfos").Count(),
+		gorethink.DB(r.ppsDbName).Table("JobInfos").Count(),
 		"Pipelines",
-		gorethink.DB(ppsDbName).Table("PipelineInfos").Count(),
-	).Run(dbClient)
+		gorethink.DB(r.ppsDbName).Table("PipelineInfos").Count(),
+	).Run(r.dbClient)
 	if err != nil {
 		protolion.Errorf("Error Fetching Metrics:%+v", err)
 	}
 	cursor.One(&metrics)
 }
 
-// ReportMetrics blocks and reports metrics, if modified, to the
-// given kubernetes client every 15 seconds.
-func ReportMetrics(clusterID string, kubeClient *kube.Client, address string, pfsDbName string, ppsDbName string) {
-	dbClient, err := db.DbConnect(address)
-	if err != nil {
-		protolion.Errorf("Error connected to DB when reporting metrics: %v\n", err)
-		return
-	}
+// ReportMetrics blocks and reports metrics every 15 seconds
+func (r *Reporter) ReportMetrics() {
+	reportingTicker := time.NewTicker(time.Second * 15)
 	for {
-		metrics := &Metrics{}
-		dbMetrics(dbClient, pfsDbName, ppsDbName, metrics)
-		externalMetrics(kubeClient, metrics)
-		metrics.ID = clusterID
-		metrics.PodID = uuid.NewWithoutDashes()
-		metrics.Version = version.PrettyPrintVersion(version.Version)
-		reportSegment(metrics)
-		<-time.After(15 * time.Second)
+		select {
+		case incrementAction := <-incrementActionChannel:
+			if userActions[incrementAction.user] == nil {
+				userActions[incrementAction.user] = make(countableActions)
+			}
+			val, ok := userActions[incrementAction.user][incrementAction.action]
+			if ok {
+				userActions[incrementAction.user][incrementAction.action] = val.(uint64) + uint64(1)
+			} else {
+				userActions[incrementAction.user][incrementAction.action] = uint64(0)
+			}
+			break
+		case batchOfActions := <-readBatchOfActionsChannel:
+			batchOfActions.channel <- userActions
+			userActions = make(countableUserActions)
+		case <-reportingTicker.C:
+			r.reportClusterMetrics()
+			r.reportUserMetrics()
+		}
```

Comment:
Those concerns makes sense

a) it would be pretty straightforward to use a goroutine for the code that's actually making the request to segment
b) since I'm resetting the counts per user (not collecting cumulative call counts), I think I do need a channel just for resetting the batch of objects ... otherwise there could easily be overcounting